<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#STAGE-II-1-CREATE-MASTER-SALES-FILE" data-toc-modified-id="STAGE-II-1-CREATE-MASTER-SALES-FILE-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>STAGE II-1 CREATE MASTER SALES FILE</a></span></li><li><span><a href="#STAGE-II-2-CREATE-MASTER-RENT-FILE" data-toc-modified-id="STAGE-II-2-CREATE-MASTER-RENT-FILE-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>STAGE II-2 CREATE MASTER RENT FILE</a></span></li><li><span><a href="#STAGE-II-3-MERGE-MASTER-SALES-AND-RENT-DATA" data-toc-modified-id="STAGE-II-3-MERGE-MASTER-SALES-AND-RENT-DATA-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>STAGE II-3 MERGE MASTER SALES AND RENT DATA</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#@all:-do-we-just-drop-the-6-nulls-in-rent?" data-toc-modified-id="@all:-do-we-just-drop-the-6-nulls-in-rent?-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>@all: do we just drop the 6 nulls in rent?</a></span></li></ul></li></ul></li><li><span><a href="#The-'Steve'-DF" data-toc-modified-id="The-'Steve'-DF-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>The 'Steve' DF</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#NOTE:-@All" data-toc-modified-id="NOTE:-@All-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span>NOTE: @All</a></span></li></ul></li></ul></li><li><span><a href="#BELOW-ARE-CODES-FOR-STAGE-I-CLEANING" data-toc-modified-id="BELOW-ARE-CODES-FOR-STAGE-I-CLEANING-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>BELOW ARE CODES FOR STAGE I CLEANING</a></span><ul class="toc-item"><li><span><a href="#STEP1.-CLEAN-SALES-DATA-(SINGLE-QUARTER)" data-toc-modified-id="STEP1.-CLEAN-SALES-DATA-(SINGLE-QUARTER)-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>STEP1. CLEAN SALES DATA (SINGLE QUARTER)</a></span><ul class="toc-item"><li><span><a href="#End-of-the-sales-data-cleaning-process" data-toc-modified-id="End-of-the-sales-data-cleaning-process-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>End of the sales data cleaning process</a></span></li></ul></li></ul></li><li><span><a href="#STEP2.-CLEAN-RENT-DATA-(SINGLE-QUARTER)" data-toc-modified-id="STEP2.-CLEAN-RENT-DATA-(SINGLE-QUARTER)-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>STEP2. CLEAN RENT DATA (SINGLE QUARTER)</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#End-of-the-rent-data-cleaning-process" data-toc-modified-id="End-of-the-rent-data-cleaning-process-6.0.1"><span class="toc-item-num">6.0.1&nbsp;&nbsp;</span>End of the rent data cleaning process</a></span></li></ul></li></ul></li><li><span><a href="#STEP3.-JOIN-SALES-AND-RENT-DATA-OF-THE-SAME-QUARTER" data-toc-modified-id="STEP3.-JOIN-SALES-AND-RENT-DATA-OF-THE-SAME-QUARTER-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>STEP3. JOIN SALES AND RENT DATA OF THE SAME QUARTER</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#!!-Please-note-!!" data-toc-modified-id="!!-Please-note-!!-7.0.1"><span class="toc-item-num">7.0.1&nbsp;&nbsp;</span>!! Please note !!</a></span></li></ul></li></ul></li><li><span><a href="#STEP4-Repeat-above-process-to-the-sales-and-rent-data-of-the-rest-quarters" data-toc-modified-id="STEP4-Repeat-above-process-to-the-sales-and-rent-data-of-the-rest-quarters-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>STEP4 Repeat above process to the sales and rent data of the rest quarters</a></span></li><li><span><a href="#STEP5-Concat-merged-sales-and-rent-quarterly-df-into-one-master-df" data-toc-modified-id="STEP5-Concat-merged-sales-and-rent-quarterly-df-into-one-master-df-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>STEP5 Concat merged sales and rent quarterly df into one master df</a></span></li><li><span><a href="#STEP-6-Add-postcode-to-LGA-Mapping-to-the-master-df" data-toc-modified-id="STEP-6-Add-postcode-to-LGA-Mapping-to-the-master-df-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>STEP 6 Add postcode to LGA Mapping to the master df</a></span></li><li><span><a href="#STEP-7-EDA-for-Interest-Rates-and-Merge-to-Master-DF" data-toc-modified-id="STEP-7-EDA-for-Interest-Rates-and-Merge-to-Master-DF-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>STEP 7 EDA for Interest Rates and Merge to Master DF</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#@all-which-df-should-be-used-for-merging?" data-toc-modified-id="@all-which-df-should-be-used-for-merging?-11.0.1"><span class="toc-item-num">11.0.1&nbsp;&nbsp;</span>@all which df should be used for merging?</a></span></li></ul></li></ul></li><li><span><a href="#STEP-8-EDA-for-Bond-Yield-Data-and-Merge-to-Master-DF" data-toc-modified-id="STEP-8-EDA-for-Bond-Yield-Data-and-Merge-to-Master-DF-12"><span class="toc-item-num">12&nbsp;&nbsp;</span>STEP 8 EDA for Bond Yield Data and Merge to Master DF</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#@all-which-dataframe-should-be-used-for-merging?" data-toc-modified-id="@all-which-dataframe-should-be-used-for-merging?-12.0.1"><span class="toc-item-num">12.0.1&nbsp;&nbsp;</span>@all which dataframe should be used for merging?</a></span></li></ul></li></ul></li></ul></div>

In [2]:
import pandas as pd
import numpy as np 

# STAGE II-1 CREATE MASTER SALES FILE

In [ ]:
# Read sales 2017-2018

sales17_18 = "Files/sales_2017_2018"
sales17_18 = pd.read_csv(sales17_18, usecols=['postcode', 'dwelling_type', 'median_price', 'mean_price',
                                               'sales_no', 'Qdelta_median', 'Adelta_median', 'Qdelta_count',
                                               'Adelta_count', 'key', 'time_period', 'year', 'quarter'])

cols = ['postcode', 'key', 'time_period', 'year', 'quarter', 
        'dwelling_type', 'median_price', 'mean_price','sales_no', 
        'Qdelta_median', 'Adelta_median', 'Qdelta_count','Adelta_count' ]

sales17_18 = sales17_18[cols]
sales17_18.head(1)

In [ ]:
# Read sales 2019-2021

sales19_21 = "Files/sales_2019_2021"
sales19_21 = pd.read_csv(sales19_21, usecols=['postcode', 'dwelling_type', 'median_price', 'mean_price',
                                               'sales_no', 'Qdelta_median', 'Adelta_median', 'Qdelta_count',
                                               'Adelta_count', 'key', 'time_period', 'year', 'quarter'])

cols = ['postcode', 'key', 'time_period', 'year', 'quarter', 
        'dwelling_type', 'median_price', 'mean_price','sales_no', 
        'Qdelta_median', 'Adelta_median', 'Qdelta_count','Adelta_count' ]

sales19_21 = sales19_21[cols]
sales19_21.head(1)

In [ ]:
# Concatenate the two sales files 
sales_full = pd.concat([sales17_18, sales19_21])

# Check if all quarters are present
print(sales_full.groupby('time_period').size())

In [ ]:
# Check null values
sales_full.isnull().sum()

# We don't really care about null values in the delta columns right?

In [ ]:
sales_full.head()

In [ ]:
sales_full.to_csv('Files/Cleaned/Sales_2017Q3_2021Q1_Clean.csv', index=False)

# STAGE II-2 CREATE MASTER RENT FILE

In [ ]:
# Read rent 2017-2018
rent17_18 = "Files/rent_2017_2018"
rent17_18 = pd.read_csv(rent17_18)

# Read rent 2019-2021
rent19_21 = "Files/rent_2019_2021"
rent19_21 = pd.read_csv(rent19_21)

# Concat both rent files
rent_full = pd.concat([rent17_18, rent19_21])
rent_full.head()

In [ ]:
# Check null values
rent_full.isnull().sum()

In [ ]:
# Check all quarters are presented
rent_full.groupby('key').size()

**NOTE:** rkey = skey-1 for the same quarter

In [ ]:
# Need to map keys to time_periods so that can merge it with sales data

tp = ['2017 Q3', '2017 Q4', 
      '2018 Q1', '2018 Q2', '2018 Q3', '2018 Q4', 
      '2019 Q1', '2019 Q2', '2019 Q3', '2019 Q4', 
      '2020 Q1', '2020 Q2', '2020 Q3', '2020 Q4', 
      '2021 Q1']
rkeys = ['r121','r122',
         'r123','r124','r125','r126',
         'r127','r128','r129','r130',
         'r131','r132','r133','r134',
         'r135']

rent_full['time_period'] = np.nan

for i in list(range(0,15)):
    rent_full.loc[rent_full['key']==rkeys[i], 'time_period']=tp[i]
    
rent_full.groupby('time_period').size()

In [ ]:
# Update column name of 'key'
rent_full = rent_full.rename(columns={'key':'rkey'})


# Change columns order
cols = ['postcode', 'rkey', 'time_period', 'median_rent_newb', 'new_bonds_no', 'total_bonds_no',
       'Qdelta_median_rent', 'Qdelta_new_bonds', 'Adelta_median_rent',
       'Adelta_new_bonds']
rent_full = rent_full[cols]

rent_full.head()

In [ ]:
# Save full rent data into csv
rent_full.to_csv('Files/Cleaned/Rent_2017Q4_2021Q2_Clean.csv', index=False)

# STAGE II-3 MERGE MASTER SALES AND RENT DATA

In [ ]:
# Change the name of 'key' column from sales to 'skey' to differentiate from 'rkey' in rent
sales_full = sales_full.rename(columns={'key':'skey'})

In [ ]:
sales_rent_full = pd.merge(sales_full, rent_full, how='left',
                           left_on=['postcode','time_period'],
                           right_on=['postcode', 'time_period'])

sales_rent_full.head()

In [ ]:
print(sales_rent_full.shape)
print(sales_rent_full.groupby('time_period').size())

In [ ]:
print(sales_rent_full.isnull().sum())

### @all: do we just drop the 6 nulls in rent?
I'd say yes since there are only six values (Felix)

In [ ]:
# Check the 6 postcodes that are null in rent:
sales_rent_full.loc[sales_rent_full['rkey'].isnull()]

In [ ]:
sales_rent_full.to_csv('Files/Cleaned/Master_Sales_Rent_2017Q4_2021Q1.csv', index=False)

# The 'Steve' DF

In [ ]:
# Create a subset that only contains 5 Quarters of data (FOR TESTING PURPOSE)
subset = sales_rent_full.loc[sales_rent_full['time_period'].isin(['2020 Q1','2020 Q2','2020 Q3','2020 Q4','2021 Q1'])]

# Get some of the (potentially) unnecessary variables
subset = subset.drop(columns=['Qdelta_median','Adelta_median','Qdelta_count','Adelta_count',
                              'Qdelta_median_rent', 'Adelta_median_rent','Qdelta_new_bonds','Adelta_new_bonds'],
                     axis=1)

# And only keep 'Total' dwelling type (i.e. get rid of Strata and Non-strata)
subset_total = subset.loc[subset['dwelling_type'] == 'Total']

print(subset_total.groupby('time_period').size(),'\n')
print(subset_total.groupby('dwelling_type').size(),'\n')
print(subset_total.shape)

In [ ]:
print(subset_total.isnull().sum())

In [ ]:
print("number of postcode in Q4 2020:",
      subset_total.loc[subset_total['time_period']=='2020 Q4', 'postcode'].nunique())
print("number of postcode in Q1 2021:",
      subset_total.loc[subset_total['time_period']=='2021 Q1', 'postcode'].nunique())

### NOTE:

Because each quarterly table has different numbers of postcodes and they're not an exact match from quarter to quarter, there will be additional null values generated after pivoting simply due to certain postcodes having data in some quarters but not in others. We have two options:

**OPTION1: Impute na with mean or median** 

In [ ]:
pivot = subset_total.pivot_table(index='postcode', columns='time_period', values=
                                 ['median_price', 'mean_price', 'sales_no', 
                                  'median_rent_newb','new_bonds_no', 'total_bonds_no']).round(2)

pivot.head()

In [ ]:
pivot.columns = [' '.join(col) for col in pivot.columns]
pivot.head()

In [ ]:
print(pivot.shape)
print(pivot.isnull().sum())

In [ ]:
pivot_fillna = pivot.fillna(pivot.median()).round(1)
pivot_fillna.isnull().sum()

In [ ]:
pivot_fillna.to_csv('Files/Cleaned/Pivot_Sales_Rent_5Quarters_Imputed.csv', index=False)

**OPTION2: only keep postcodes that are present in all quarters:**

In [ ]:
skeys = list(sales_full['skey'].unique())

D={}
for key in skeys:
    D[str(key)]= sorted(list(sales_full.loc[sales_full['skey']==key, 'postcode'].unique()))
    
for key in skeys:
    print("postcodes in", key,":", len(D[key]))

In [ ]:
ls122 = D['s122']
ls123 = D['s123']
ls123 = D['s123']
ls124 = D['s124']
ls125 = D['s125']
ls126 = D['s126']
ls127 = D['s127']
ls128 = D['s128']
ls129 = D['s129']
ls130 = D['s130']
ls131 = D['s131']
ls132 = D['s132']
ls133 = D['s133']
ls134 = D['s134']
ls135 = D['s135']
ls136 = D['s136']

shared_postcodes = list(set.intersection(*map(set,[ls122, ls123,ls124,ls125,ls126,ls127,ls128,ls129,ls130,
                                                   ls131,ls132,ls133,ls134,ls135,ls136])))
shared_postcodes = sorted(shared_postcodes)
print(len(shared_postcodes))

In [ ]:
# For the purpose of the test DF (5 quarters) - find shared postcodes of this 5 quarters
shared_pc_5 = list(set.intersection(*map(set,[ls132,ls133,ls134,ls135,ls136])))
print("postcodes in all 5 quarters:", len(shared_pc_5))

In [ ]:
# Keep only this 493 postcodes
subset_total = subset_total.loc[subset_total['postcode'].isin(shared_pc_5)]
subset_total['postcode'].nunique()

In [ ]:
# Then pivot

pivot_493 = subset_total.pivot_table(index='postcode', columns='time_period', values=
                                     ['median_price', 'mean_price', 'sales_no', 'median_rent_newb','new_bonds_no', 'total_bonds_no']).round(2)

print(pivot_493.shape)

In [ ]:
pivot_493.columns = [' '.join(col) for col in pivot_493.columns]
pivot_493.head()

In [ ]:
pivot_493.isnull().sum()

In [ ]:
pivot_493.to_csv('Files/Cleaned/Pivot_Sales_Rent_5Quarters_SharedPOA.csv', index=False)


<BR>

# BELOW ARE CODES FOR STAGE I CLEANING
______

## STEP1. CLEAN SALES DATA (SINGLE QUARTER) #

In [3]:
# Read one file at a time into df, name df by 's'+'issue number ''
s136 = "Files/Sales/2019_2021/Issue-136-Sales-tables-March-2021-quarter.xlsx"
s136 = pd.read_excel(s136, sheet_name="Postcode", na_values='-', header=6)

In [4]:
# Rename columns
rename_cols= {'Postcode':'postcode', 
             'Dwelling Type':'dwelling_type', 
             "First Quartile Sales Price\n$'000s" : '25%_price',
             "Median Sales Price\n$'000s" : 'median_price', 
             "Third Quartile Sales Price\n'000s" : '75%_price',
             "Mean Sales Price\n$'000s" : 'mean_price',
             'Sales\nNo.':'sales_no',
             'Qtly change in Median':'Qdelta_median_price',
             'Annual change in Median':'Adelta_median_price',
             'Qtly change in Count':'Qdelta_sales_no',
             'Annual change in Count':'Adelta_sales_no'}
s136.rename(columns=rename_cols, inplace=True)

# Drop unwanted columns
s136 = s136.drop(columns=['25%_price', '75%_price'], axis=1)

s136.head(5)

postcode dwelling_type  median_price  mean_price sales_no  \
0      2000         Total        1371.0      2794.0      184   
1      2000        Strata        1371.0      2794.0      184   
2      2007         Total         763.0       754.0        s   
3      2007    Non Strata           NaN         NaN      NaN   
4      2007        Strata         710.0       688.0        s   

   Qdelta_median_price  Adelta_median_price  Qdelta_sales_no  Adelta_sales_no  
0               0.1425               0.0711           0.0888           0.5862  
1               0.1331               0.0632           0.1018           0.6429  
2               0.1713               0.0235           0.0000           0.3000  
3                  NaN                  NaN              NaN              NaN  
4               0.0906              -0.0470          -0.0769           0.2000

In [5]:
print("number of postcodes:", s136['postcode'].nunique())
print("number of rows:", s136.shape[0],'\n')
print(s136.isnull().sum())

number of postcodes: 544
number of rows: 1427 

postcode                 0
dwelling_type            0
median_price           316
mean_price             316
sales_no               316
Qdelta_median_price    316
Adelta_median_price    321
Qdelta_sales_no        316
Adelta_sales_no        321
dtype: int64


<b>IMPUTATION</b>

Roughly 1/4 of the data contains null values, which comes from:
<em>"for confidentiality, we don't report sales in any geographical area where the number of sales is 10 or less (shown as na).</em> Also: <em>"Statistics calculated from sample sizes between 10 an 30 are shown by an 's' in the relevant table".</em> This will be the primary rules of imputation:
* **For 'sales_no':**
    * Replace na with 5
    * Replace s with 20
    
* **For 'median_price' and 'mean_price'**
    * Fill na with the median of price of the same dwelling type

In [6]:
# Imputation of sales number

# na
s136.loc[s136['sales_no'].isnull(), 'sales_no'] = 5.0
s136['sales_no'].isnull().any()

# s
s136.loc[s136['sales_no'] == 's', 'sales_no'] = 20.0
s136['sales_no'] = s136['sales_no'].astype(float) # Cast type as float

print("number of na in sales_no:", s136['sales_no'].isnull().sum())
print("data type of sales_no:", s136['sales_no'].dtype)

number of na in sales_no: 0
data type of sales_no: float64


In [7]:
# Imputation of median_price (by dwelling type)

# Total
imp_total = s136.loc[(s136['median_price'].notna()) & (s136['dwelling_type']=='Total'),
                     'median_price'].median() # calculate imputer value 
    
s136.loc[(s136['median_price'].isnull()) & (s136['dwelling_type']=='Total'),
         'median_price']= imp_total #impute

# Strata 
imp_strata = s136.loc[(s136['median_price'].notna()) & (s136['dwelling_type']=='Strata'),
                     'median_price'].median() 
    
s136.loc[(s136['median_price'].isnull()) & (s136['dwelling_type']=='Strata'),
         'median_price']= imp_strata 

# Non-Strata
imp_ns = s136.loc[(s136['median_price'].notna()) & (s136['dwelling_type']=='Non Strata'),
                     'median_price'].median() 
    
s136.loc[(s136['median_price'].isnull()) & (s136['dwelling_type']=='Non Strata'),
         'median_price']= imp_ns

print("number of na in median_price:", s136['median_price'].isnull().sum())

number of na in median_price: 0


In [8]:
# Imputation of mean_price (by dwelling type)

# Total
imp_total = s136.loc[(s136['mean_price'].notna()) & (s136['dwelling_type']=='Total'),
                     'mean_price'].median() # calculate imputer value 
    
s136.loc[(s136['mean_price'].isnull()) & (s136['dwelling_type']=='Total'),
         'mean_price']= imp_total #impute

# Strata 
imp_strata = s136.loc[(s136['mean_price'].notna()) & (s136['dwelling_type']=='Strata'),
                     'mean_price'].median() 
    
s136.loc[(s136['mean_price'].isnull()) & (s136['dwelling_type']=='Strata'),
         'mean_price']= imp_strata 

# Non-Strata
imp_ns = s136.loc[(s136['mean_price'].notna()) & (s136['dwelling_type']=='Non Strata'),
                  'mean_price'].median() 
    
s136.loc[(s136['mean_price'].isnull()) & (s136['dwelling_type']=='Non Strata'),
         'mean_price']= imp_ns

print("number of na in mean_price:", s136['mean_price'].isnull().sum())

number of na in mean_price: 0


In [9]:
s136.describe().round(2)

postcode  median_price  mean_price  sales_no  Qdelta_median_price  \
count   1427.00       1427.00     1427.00   1427.00              1111.00   
mean    2378.20        923.04      994.20     51.47                 0.06   
std      256.21        628.89      706.76     58.94                 0.14   
min     2000.00        120.00      125.00      5.00                -0.90   
25%     2148.00        610.00      635.50     20.00                -0.01   
50%     2330.00        760.50      801.50     20.00                 0.05   
75%     2575.00        960.00     1072.50     68.00                 0.12   
max     2880.00       6400.00     7374.00    397.00                 0.78   

       Adelta_median_price  Qdelta_sales_no  Adelta_sales_no  
count              1106.00          1111.00          1106.00  
mean                  0.13            -0.03             0.62  
std                   0.20             0.41             0.85  
min                  -0.46            -0.59            -0.49  
25%                   0.03            -0.22             0.17  
50%                   0.11            -0.10             0.41  
75%                   0.20             0.06             0.74  
max                   2.04             6.50             7.00

**ADD TIME PERIOD TAG**

In [10]:
s136['key_s'] = 's136'
s136['time_period'] = '2021 Q1'
s136['year'] = '2021'
s136['quarter'] = '1'

print(s136.shape)

(1427, 13)


In [11]:
s136

postcode dwelling_type  median_price  mean_price  sales_no  \
0         2000         Total        1371.0      2794.0     184.0   
1         2000        Strata        1371.0      2794.0     184.0   
2         2007         Total         763.0       754.0      20.0   
3         2007    Non Strata         843.0       888.0       5.0   
4         2007        Strata         710.0       688.0      20.0   
...        ...           ...           ...         ...       ...   
1422      2877    Non Strata         195.0       212.0      20.0   
1423      2879         Total         760.5       801.5       5.0   
1424      2879    Non Strata         843.0       888.0       5.0   
1425      2880         Total         120.0       146.0     107.0   
1426      2880    Non Strata         120.0       146.0     107.0   

      Qdelta_median_price  Adelta_median_price  Qdelta_sales_no  \
0                  0.1425               0.0711           0.0888   
1                  0.1331               0.0632           0.1018   
2                  0.1713               0.0235           0.0000   
3                     NaN                  NaN              NaN   
4                  0.0906              -0.0470          -0.0769   
...                   ...                  ...              ...   
1422              -0.0714               0.3636           0.5385   
1423                  NaN                  NaN              NaN   
1424                  NaN                  NaN              NaN   
1425              -0.0323              -0.0400           0.1146   
1426              -0.0323              -0.0400           0.1146   

      Adelta_sales_no key_s time_period  year quarter  
0              0.5862  s136     2021 Q1  2021       1  
1              0.6429  s136     2021 Q1  2021       1  
2              0.3000  s136     2021 Q1  2021       1  
3                 NaN  s136     2021 Q1  2021       1  
4              0.2000  s136     2021 Q1  2021       1  
...               ...   ...         ...   ...     ...  
1422           4.0000  s136     2021 Q1  2021       1  
1423              NaN  s136     2021 Q1  2021       1  
1424              NaN  s136     2021 Q1  2021       1  
1425           0.5070  s136     2021 Q1  2021       1  
1426           0.5070  s136     2021 Q1  2021       1  

[1427 rows x 13 columns]

### End of the sales data cleaning process ###
----

# STEP2. CLEAN RENT DATA (SINGLE QUARTER) #


In [12]:
# Read into df
r135 = "Files/Rent/2019_2021/Issue-135-Rent-tables-March-2021-quarter.xlsx"
r135 = pd.read_excel(r135, sheet_name="Postcode", na_values='-', header=7)

# Drop unwanted columns
r135 = r135.drop(columns=['First Quartile Weekly Rent for New Bonds\n$',
                          'Third Quartile Weekly Rent for New Bonds\n$'],
                axis=1)

# Rename columns
rename_cols= {'Postcode':'postcode',
              'Dwelling Types':'dwelling_type', 
              'Number of Bedrooms':'bed_number',
              'Median Weekly Rent for New Bonds\n$': 'median_rent_newb',
              'New Bonds Lodged\nNo.' : 'new_bonds_no',
              'Total Bonds Held\nNo.': 'total_bonds_no',
              'Quarterly change in Median Weekly Rent':'Qdelta_median_rent',
              'Annual change in Median Weekly Rent':'Adelta_median_rent',
              'Quarterly change in New Bonds Lodged':'Qdelta_new_bonds',
              'Annual change in New Bonds Lodged':'Adelta_new_bonds'}
r135.rename(columns=rename_cols,inplace=True)

r135.head(10)

postcode dwelling_type          bed_number  median_rent_newb new_bonds_no  \
0      2000         Total               Total             600.0         1469   
1      2000         Total           Bedsitter             365.0           89   
2      2000         Total           1 Bedroom             540.0          741   
3      2000         Total       Not Specified             445.0           34   
4      2000         Total          2 Bedrooms             750.0          517   
5      2000         Total          3 Bedrooms            1000.0           74   
6      2000         Total  4 or more Bedrooms            1675.0            s   
7      2000         House               Total             710.0            s   
8      2000         House           Bedsitter               NaN          NaN   
9      2000         House           1 Bedroom             500.0            s   

  total_bonds_no  Qdelta_median_rent  Adelta_median_rent  Qdelta_new_bonds  \
0           9327              0.0909             -0.1429           -0.1384   
1            382              0.0429             -0.2843            0.0349   
2           4063              0.1020             -0.1692           -0.1136   
3            511             -0.1524             -0.3904           -0.3200   
4           3741              0.1194             -0.1979           -0.2083   
5            575              0.1111             -0.2780            0.0137   
6             55              0.6919             -0.0429            1.0000   
7            184              0.0965             -0.1647           -0.3182   
8            NaN                 NaN                 NaN               NaN   
9             55              0.0000             -0.2187           -0.2143   

   Adelta_new_bonds  
0            0.1943  
1            0.9778  
2           -0.0326  
3            0.2593  
4            0.6056  
5            0.0882  
6            6.0000  
7            0.7647  
8               NaN  
9            0.3750

In [13]:
# Check df shape and null values
print(r135.shape)
print(r135.isnull().sum())

(16397, 10)
postcode                  0
dwelling_type             0
bed_number                0
median_rent_newb      12274
new_bonds_no          12274
total_bonds_no         6095
Qdelta_median_rent    12276
Adelta_median_rent    12277
Qdelta_new_bonds      12276
Adelta_new_bonds      12277
dtype: int64


**NOTE:**

Note that an alarming 3/4 of the data has null values. This is because the data is broken down to very granular level - first by dwelling type (Total, house, townhouse, flat/unit, other) and then by bed_numbers (see below cell).

In [14]:
print(r135.groupby('dwelling_type').size(),'\n')
print(r135.groupby('bed_number').size())

dwelling_type
Flat/Unit    3160
House        3653
Other        3257
Total        3903
Townhouse    2424
dtype: int64 

bed_number
1 Bedroom             2336
2 Bedrooms            2655
3 Bedrooms            2637
4 or more Bedrooms    2330
Bedsitter             1296
Not Specified         2310
Total                 2833
dtype: int64


**AGGREGATION:**

Given the amount of na present, we'll **aggregate the data to the total level of each postcode** (i.e. only consider the total number of bonds without classifying them into dwelling types nor bedroom numbers) to avoid excessive imputation. Although we'll inevitably lose the richness of information, the main function of the rent data is to serve as an input variable in the prediction of sales, where this level of granularity is not strictly necessary.

In [15]:
# Aggregate dwelling type and bed number, save as new df r135_ag

r135_ag = r135.loc[(r135['bed_number']=='Total') & (r135['dwelling_type']=='Total')]
r135_ag = r135_ag.drop(columns=['bed_number','dwelling_type'], axis=1) # Drop bed_number and dwelling_type

print(r135_ag.shape)
print(r135_ag.isnull().sum())

(617, 8)
postcode                0
median_rent_newb      181
new_bonds_no          181
total_bonds_no         51
Qdelta_median_rent    181
Adelta_median_rent    181
Qdelta_new_bonds      181
Adelta_new_bonds      181
dtype: int64


By aggregating the data, we're able to bring down the proportion of na from 3/4 to around 1/3. But there's still need for imputation. According to the data interpretation note:

<em><b>"For confidentiality, we don't report rents in any geographical area where the number of new bonds is 10 or less (shown as na). Statistics calculated from sample sizes between 10 an 30 are shown by an 's' in the relevant table"</b></em>

In [16]:
print("number of 's' in new_bonds_no:", r135_ag.loc[r135_ag['new_bonds_no']=='s'].shape[0])
print("number of 's' in total_bonds_noA:", r135_ag.loc[r135_ag['total_bonds_no']=='s'].shape[0],"\n")
print("number of na in new_bonds_no:", r135_ag.loc[r135_ag['new_bonds_no'].isnull()].shape[0])
print("number of na in total_bonds_no:", r135_ag.loc[r135_ag['total_bonds_no'].isnull()].shape[0])

number of 's' in new_bonds_no: 88
number of 's' in total_bonds_noA: 47 

number of na in new_bonds_no: 181
number of na in total_bonds_no: 51


<b>IMPUTATION</b>
* For 'new_bonds_no' and 'total_bonds_no' columns:
    * Impute na with 5
    * Impute s with 20
    
* For 'median_rent_newb' column
    * Impute na with median of rents of all POAs

In [17]:
# Impute 's' in 'new_bonds_no' and 'total_bonds_no' with 20
r135_ag.loc[r135_ag['new_bonds_no']=='s','new_bonds_no'] = 20.0
r135_ag.loc[r135_ag['total_bonds_no']=='s', 'total_bonds_no'] = 20.0

# Impute na in 'new_bonds_no' and 'total_bonds_no' with 5
r135_ag.loc[r135_ag['new_bonds_no'].isnull(),'new_bonds_no'] = 5.0
r135_ag.loc[r135_ag['total_bonds_no'].isnull(), 'total_bonds_no'] = 5.0

# Cast both variables as float (was object)
r135_ag['new_bonds_no'] = r135_ag['new_bonds_no'].astype(float)
r135_ag['total_bonds_no'] = r135_ag['total_bonds_no'].astype(float)

In [18]:
# Impute na in 'median_rent' with median of the column
r135_ag['median_rent_newb'].fillna(r135_ag['median_rent_newb'].median(), inplace=True)

# Check na in the df again
print(r135_ag.isnull().sum())

postcode                0
median_rent_newb        0
new_bonds_no            0
total_bonds_no          0
Qdelta_median_rent    181
Adelta_median_rent    181
Qdelta_new_bonds      181
Adelta_new_bonds      181
dtype: int64


In [19]:
# Prepare for merging
r135_ag['key_r'] = 'r135' # Add key

# Set postcode as index
r135_ag = r135_ag.set_index('postcode')
r135_ag.head(1)

median_rent_newb  new_bonds_no  total_bonds_no  Qdelta_median_rent  \
postcode                                                                       
2000                 600.0        1469.0          9327.0              0.0909   

          Adelta_median_rent  Qdelta_new_bonds  Adelta_new_bonds key_r  
postcode                                                                
2000                 -0.1429           -0.1384            0.1943  r135

In [20]:
r135_ag

median_rent_newb  new_bonds_no  total_bonds_no  Qdelta_median_rent  \
postcode                                                                       
2000                 600.0        1469.0          9327.0              0.0909   
2007                 455.0         388.0          2024.0             -0.0521   
2008                 500.0         987.0          4228.0              0.0204   
2009                 600.0         471.0          3333.0              0.0909   
2010                 525.0        1139.0          8626.0              0.0145   
...                    ...           ...             ...                 ...   
2877                 220.0          20.0           204.0             -0.0222   
2878                 455.0           5.0            20.0                 NaN   
2879                 455.0           5.0             5.0                 NaN   
2880                 270.0         124.0          1667.0              0.0800   
2905                 455.0           5.0             5.0                 NaN   

          Adelta_median_rent  Qdelta_new_bonds  Adelta_new_bonds key_r  
postcode                                                                
2000                 -0.1429           -0.1384            0.1943  r135  
2007                 -0.1727            0.1182           -0.0673  r135  
2008                 -0.1597            0.0787            0.0313  r135  
2009                 -0.1429           -0.1326            0.6469  r135  
2010                 -0.1797           -0.0164            0.2448  r135  
...                      ...               ...               ...   ...  
2877                 -0.1200            1.0000           -0.3333  r135  
2878                     NaN               NaN               NaN  r135  
2879                     NaN               NaN               NaN  r135  
2880                  0.0385            0.0081           -0.1206  r135  
2905                     NaN               NaN               NaN  r135  

[617 rows x 8 columns]

### End of the rent data cleaning process ##
----



# STEP3. JOIN SALES AND RENT DATA OF THE SAME QUARTER #

### !! Please note !! ###
**A. the issue number of the sales and rent table of the same quarter are different**
<br>For example, for Q1 2021, the issue number of the sales table is 136 and that of the rent table is 135. The resulting data frames are s136 and r135_ag. For the same quarter, the issue number of sales table is always that of the rent+1 </br>

**B. the sales table and the rent table contains different numbers of postcodes**
<br>For example, s136 (sales table) contains 544 unique postcodes while r135 (rent table) contains 617. We will merge the rent table into the sales table, i.e. only keep postcodes that are available in the sales table </br>

In [21]:
# Merge the rent df into the sales df of the same quarter (Q1 2021)
q121 = s136.join(r135_ag, on='postcode')
q121.head(1)

# Note: naming convention of merged quarterly sales+rent df
# quarter in lowercase + 2digit year
# e.g. Q4 2018 will be q418

postcode dwelling_type  median_price  mean_price  sales_no  \
0      2000         Total        1371.0      2794.0     184.0   

   Qdelta_median_price  Adelta_median_price  Qdelta_sales_no  Adelta_sales_no  \
0               0.1425               0.0711           0.0888           0.5862   

  key_s  ...  year quarter median_rent_newb  new_bonds_no  total_bonds_no  \
0  s136  ...  2021       1            600.0        1469.0          9327.0   

   Qdelta_median_rent  Adelta_median_rent  Qdelta_new_bonds  Adelta_new_bonds  \
0              0.0909             -0.1429           -0.1384            0.1943   

   key_r  
0   r135  

[1 rows x 21 columns]

In [22]:
q121['postcode'].nunique()

544

# STEP4 Repeat above process to the sales and rent data of the rest quarters #

**@Chris** - I'm assuming that we will be able to do this easily with the functions created from step 1-3?

# STEP5 Concat merged sales and rent quarterly df into one master df #

In [23]:
# NOTE: Dummy code here 
frames = [q118, q218, q318, q418, 
          q119, q219, q319, q419,
          q120, q220, q320, q420,
          q121]
sr_master = pd.concat(frames)

NameError: name 'q118' is not defined

# STEP 6 Add postcode to LGA Mapping to the master df #

------

# 07/OCT/2021 Alexis worked

In [24]:
# Number of postcodes in the master datafile

master = pd.read_csv("Files/Cleaned/Master_Sales_Rent_2017Q4_2021Q1.csv")

print("number of postcodes in the master sales df:", master['postcode'].nunique())
pc_master = master['postcode'].unique()

number of postcodes in the master sales df: 587


In [25]:
# Read the mapping data into df
lga_poa = "Files/Area/Postcode_and_LGA.xlsx"
mapping = pd.read_excel(lga_poa, sheet_name="SuburbLGA", 
                        usecols=['lganame','councilnam','suburbname','postcode'])

# Rename columns 
rename_cols= {'lganame':'LGA',
              'councilnam':'council', 
              'suburbname':'suburb'}
mapping.rename(columns=rename_cols,inplace=True)

mapping.head()

LGA              council           suburb  postcode
0  ALBURY CITY  ALBURY CITY COUNCIL           ALBURY    2640.0
1  ALBURY CITY  ALBURY CITY COUNCIL      EAST ALBURY    2640.0
2  ALBURY CITY  ALBURY CITY COUNCIL        ETTAMOGAH    2640.0
3  ALBURY CITY  ALBURY CITY COUNCIL          GLENROY    2640.0
4  ALBURY CITY  ALBURY CITY COUNCIL  HAMILTON VALLEY    2641.0

In [26]:
# Check null values in mapping
print(mapping.isnull().sum())

LGA         12
council     12
suburb       0
postcode    12
dtype: int64


In [27]:
# Drop na
mapping = mapping.dropna()

In [28]:
print("number of postcodes in the mapping df:", mapping['postcode'].nunique())
pc_map = mapping['postcode'].unique()

number of postcodes in the mapping df: 622


In [29]:
pc_shared = list(set(pc_master).intersection(pc_map))
print("number of postcodes in both lists:", len(pc_shared))

number of postcodes in both lists: 587


**GREAT NEWS - ALL POSTCODES IN THE MASTER SALES/RENT DF CAN BE FOUND IN THE MAPPING FILE**

The next step is to sort out the mapping df and join it with the master df.

In [30]:
# Dealing with the unincorporated

unico = mapping.loc[mapping.LGA=='UNINCORPORATED']
print("number of unicorporated postcodes:", unico['postcode'].nunique())
unico = unico.sort_values(by='postcode', ascending=True)
unico

number of unicorporated postcodes: 20


LGA                  council           suburb  postcode
4576  UNINCORPORATED           UNINCORPORATED       BARANGAROO    2000.0
4581  UNINCORPORATED           UNINCORPORATED      DAWES POINT    2000.0
4605  UNINCORPORATED           UNINCORPORATED           SYDNEY    2000.0
4594  UNINCORPORATED           UNINCORPORATED    MILLERS POINT    2000.0
4602  UNINCORPORATED           UNINCORPORATED          PYRMONT    2009.0
4611  UNINCORPORATED           UNINCORPORATED    WOOLLOOMOOLOO    2011.0
4582  UNINCORPORATED           UNINCORPORATED       DOUBLE BAY    2028.0
4609  UNINCORPORATED           UNINCORPORATED      WATSONS BAY    2030.0
4607  UNINCORPORATED           UNINCORPORATED         VAUCLUSE    2030.0
4575  UNINCORPORATED           UNINCORPORATED          BALMAIN    2041.0
4577  UNINCORPORATED           UNINCORPORATED       BIRCHGROVE    2041.0
4574  UNINCORPORATED           UNINCORPORATED       ABBOTSFORD    2046.0
4583  UNINCORPORATED           UNINCORPORATED        DRUMMOYNE    2047.0
4587  UNINCORPORATED           UNINCORPORATED       KIRRIBILLI    2061.0
4589  UNINCORPORATED           UNINCORPORATED  LANE COVE NORTH    2066.0
4596  UNINCORPORATED           UNINCORPORATED           MOSMAN    2088.0
4588  UNINCORPORATED           UNINCORPORATED    KURRABA POINT    2089.0
4598  UNINCORPORATED           UNINCORPORATED      NEUTRAL BAY    2089.0
4591  UNINCORPORATED           UNINCORPORATED            MANLY    2095.0
4586  UNINCORPORATED           UNINCORPORATED     HUNTERS HILL    2110.0
4585  UNINCORPORATED           UNINCORPORATED      GLADESVILLE    2111.0
4603  UNINCORPORATED  UNINCORPORATED FAR WEST           SCOTIA    2648.0
4601  UNINCORPORATED  UNINCORPORATED FAR WEST        POONCARIE    2648.0
4610  UNINCORPORATED  UNINCORPORATED FAR WEST     WHITE CLIFFS    2836.0
4608  UNINCORPORATED  UNINCORPORATED FAR WEST        WANAARING    2840.0
4592  UNINCORPORATED  UNINCORPORATED FAR WEST         MENINDEE    2879.0
4597  UNINCORPORATED  UNINCORPORATED FAR WEST       MUTAWINTJI    2880.0
4604  UNINCORPORATED  UNINCORPORATED FAR WEST        SILVERTON    2880.0
4595  UNINCORPORATED  UNINCORPORATED FAR WEST       MILPARINKA    2880.0
4606  UNINCORPORATED  UNINCORPORATED FAR WEST       TIBOOBURRA    2880.0
4580  UNINCORPORATED  UNINCORPORATED FAR WEST   BROUGHAMS GATE    2880.0
4590  UNINCORPORATED  UNINCORPORATED FAR WEST     LITTLE TOPAR    2880.0
4579  UNINCORPORATED  UNINCORPORATED FAR WEST      BROKEN HILL    2880.0
4584  UNINCORPORATED  UNINCORPORATED FAR WEST      FOWLERS GAP    2880.0
4600  UNINCORPORATED  UNINCORPORATED FAR WEST       PACKSADDLE    2880.0

In [31]:
# import manully looked-up LGAs for unicorporated ones

patch = ["CITY OF SYDNEY",
"CITY OF SYDNEY",
"CITY OF SYDNEY",
"CITY OF SYDNEY",
"CITY OF SYDNEY",
"CITY OF SYDNEY",
"WOOLLAHRA",
"WOOLLAHRA",
"WOOLLAHRA",
"INNER WEST",
"INNER WEST",
"CANADA BAY",
"CANADA BAY",
"NORTH SYDNEY",
"LANE COVE",
"MOSMAN",
"NORTH SYDNEY",
"NORTH SYDNEY",
"NORTHERN BEACHES",
"HUNTERS HILL",
"CITY OF RYDE",
"WENTWORTH SHIRE",
"WENTWORTH SHIRE",
"CENTRAL DARLING SHIRE",
"BOURKE SHIRE",
"CENTRAL DARLING SHIRE",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST",
"UNICORPORATED FAR WEST"]

# Add patch
unico['LGA patch'] = patch

# Drop LGA
unico = unico.drop(columns=['LGA'],axis=1)

# Remane patch as the new LGA
unico = unico.rename(columns={"LGA patch":"LGA"})

# Reorder columns
cols = ['LGA', 'council', 'suburb', 'postcode']
unico = unico[cols]

unico.head()

LGA         council         suburb  postcode
4576  CITY OF SYDNEY  UNINCORPORATED     BARANGAROO    2000.0
4581  CITY OF SYDNEY  UNINCORPORATED    DAWES POINT    2000.0
4605  CITY OF SYDNEY  UNINCORPORATED         SYDNEY    2000.0
4594  CITY OF SYDNEY  UNINCORPORATED  MILLERS POINT    2000.0
4602  CITY OF SYDNEY  UNINCORPORATED        PYRMONT    2009.0

In [32]:
# Remove unicorporated from the original mapping df
mapping = mapping.loc[mapping['LGA'] != "UNINCORPORATED"]

# Append patched unico to mapping
mapping = mapping.append(unico)

# Save complete mapping df to csv
mapping.to_csv('Files/Area/Cleaned_LGA_Mapping_07102021.csv', index=False)

**UP TO THIS POINT, WE'VE SOLVED THE UNINCORPORATED ISSUES OF THE MAPPING FILE AND CAN START MERGING**

In [33]:
# Remove the suburb and council columns
mapping_reduce = mapping.drop(columns=['council','suburb'], axis=1)

# Drop duplicates
mapping_reduce = mapping_reduce.drop_duplicates()

mapping_reduce

LGA  postcode
0                ALBURY CITY    2640.0
4                ALBURY CITY    2641.0
5                ALBURY CITY    2642.0
6                ALBURY CITY    3691.0
16         ARMIDALE REGIONAL    2350.0
...                      ...       ...
4603         WENTWORTH SHIRE    2648.0
4610   CENTRAL DARLING SHIRE    2836.0
4608            BOURKE SHIRE    2840.0
4592   CENTRAL DARLING SHIRE    2879.0
4597  UNICORPORATED FAR WEST    2880.0

[1064 rows x 2 columns]

In [34]:
# Join LGA from mapping with master
master_map = master.merge(mapping_reduce, left_on='postcode', right_on='postcode')

cols = ['postcode', 'LGA','skey', 'time_period', 'year', 'quarter', 'dwelling_type',
       'median_price', 'mean_price', 'sales_no', 'Qdelta_median',
       'Adelta_median', 'Qdelta_count', 'Adelta_count', 'rkey',
       'median_rent_newb', 'new_bonds_no', 'total_bonds_no',
       'Qdelta_median_rent', 'Qdelta_new_bonds', 'Adelta_median_rent',
       'Adelta_new_bonds']

master_map = master_map[cols]
master_map.head(1)

postcode     LGA  skey time_period  year quarter dwelling_type  \
0      2000  SYDNEY  s122     2017 Q3  2017      Q3         Total   

   median_price   mean_price  sales_no  ...  Qdelta_count  Adelta_count  rkey  \
0        1350.0  1516.328059     135.0  ...        -0.325       -0.3112  r121   

   median_rent_newb new_bonds_no  total_bonds_no  Qdelta_median_rent  \
0             640.0       1169.0          7914.0                -0.2   

   Qdelta_new_bonds  Adelta_median_rent  Adelta_new_bonds  
0            0.5545                 NaN               NaN  

[1 rows x 22 columns]

In [35]:
master_map.to_csv('Files/Cleaned/Master_Sales_Rent_2017Q4_2021Q1_wLGA.csv', index=False)

In [36]:
master_map.columns

Index(['postcode', 'LGA', 'skey', 'time_period', 'year', 'quarter',
       'dwelling_type', 'median_price', 'mean_price', 'sales_no',
       'Qdelta_median', 'Adelta_median', 'Qdelta_count', 'Adelta_count',
       'rkey', 'median_rent_newb', 'new_bonds_no', 'total_bonds_no',
       'Qdelta_median_rent', 'Qdelta_new_bonds', 'Adelta_median_rent',
       'Adelta_new_bonds'],
      dtype='object')

**THE NEXT STEP WOULD BE SOLVE THE MISMATCH OF LGA NAMES BETWEEN THE MASTER DF AND THE FEATURE DF**

`hhold`

In [37]:
df_hhold_wide = "Files/df_hhold_wide_cleaned.csv"
df_hhold_wide = pd.read_csv(df_hhold_wide)

popfile = "Files/Population/2019 NSW Population Projections ASGS 2019 LGA.xlsx"
df_hhold_wide = pd.read_excel(popfile,sheet_name='LGA Household Totals',header=6,usecols="A:C",skipfooter=3)
df_hhold_wide.columns=['LGA','hhold_count_2016','hhold_count_2021']
df_hhold_wide['LGA'] = df_hhold_wide.LGA.str.split('(').str.get(0)

df_hhold_wide['hhold_count_delta'] = df_hhold_wide.hhold_count_2021 - df_hhold_wide.hhold_count_2016
df_hhold_wide['LGA'] = df_hhold_wide['LGA'].str.strip()

hhold = df_hhold_wide
hhold.head()

LGA  hhold_count_2016  hhold_count_2021  hhold_count_delta
0             Albury             21940             23227               1287
1  Armidale Regional             11755             13041               1286
2            Ballina             18178             19080                902
3          Balranald               963              1015                 52
4  Bathurst Regional             16105             17351               1246

In [38]:
print("Number of LGAs in the population df:", hhold['LGA'].nunique())
print("Number of LGAs in the mapping df:", mapping['LGA'].nunique())

Number of LGAs in the population df: 129
Number of LGAs in the mapping df: 135


In [39]:
# Find shared LGA in both 
LGA_map = list(mapping['LGA'].unique())
LGA_hhold = list(hhold['LGA'].unique())
LGA_HHOLD = list(map(str.upper, LGA_hhold)) # Convert LGA in hhold to uppercase

LGA_shared= list(set(LGA_map).intersection(LGA_HHOLD))
print(len(LGA_shared))

122


In [40]:
# Find LGA in hhold ONLY
set(LGA_HHOLD) - set(LGA_shared)

{'ALBURY',
 'LITHGOW',
 'NAMBUCCA',
 'PARRAMATTA',
 'UNINCORPORATED NSW',
 'UPPER HUNTER SHIRE',
 'WARRUMBUNGLE SHIRE'}

In [41]:
# Find LGA in mapping ONLY
set(LGA_map) - set(LGA_shared)

{'ALBURY CITY',
 'BOURKE SHIRE',
 'CENTRAL DARLING SHIRE',
 'CITY OF PARRAMATTA',
 'CITY OF RYDE',
 'CITY OF SYDNEY',
 'LITHGOW CITY',
 'LORD HOWE ISLAND - UNINCORPORATED AREA',
 'NAMBUCCA VALLEY',
 'UNICORPORATED FAR WEST',
 'UPPER HUNTER',
 'WARRUMBUNGLE',
 'WENTWORTH SHIRE'}

In [42]:
# Match name, except for unincorporated NSW

hhold.loc[hhold.LGA=='Albury', 'LGA'] = 'ALBURY CITY'
hhold.loc[hhold.LGA=='Lithgow', 'LGA'] = 'LITHGOW CITY'
hhold.loc[hhold.LGA=='Nambucca', 'LGA'] = 'NAMBUCCA VALLEY'
hhold.loc[hhold.LGA=='Parramatta', 'LGA'] = 'CITY OF PARRAMATTA'
hhold.loc[hhold.LGA=='Upper Hunter Shire', 'LGA'] = 'UPPER HUNTER'
hhold.loc[hhold.LGA=='Warrumbungle Shire', 'LGA'] = 'WARRUMBUNGLE'

LGA_hhold = list(hhold['LGA'].unique())
LGA_HHOLD = list(map(str.upper, LGA_hhold)) # Convert LGA in hhold to uppercase

LGA_shared= list(set(LGA_map).intersection(LGA_HHOLD))
print("number of LGAs matched:", len(LGA_shared))
print("number of LGAs in the hhold df:", hhold['LGA'].nunique())

number of LGAs matched: 128
number of LGAs in the hhold df: 129


In [43]:
# Find LGA in hhold ONLY
set(LGA_HHOLD) - set(LGA_shared)

{'UNINCORPORATED NSW'}

In [44]:
# making the LGAs uppercase
hhold['LGA'] = hhold['LGA'].str.upper()
hhold.head()

LGA  hhold_count_2016  hhold_count_2021  hhold_count_delta
0        ALBURY CITY             21940             23227               1287
1  ARMIDALE REGIONAL             11755             13041               1286
2            BALLINA             18178             19080                902
3          BALRANALD               963              1015                 52
4  BATHURST REGIONAL             16105             17351               1246

#### Population movement in 5 year period
`move`

In [45]:
popfile = "Files/Population/2019 NSW Population Projections ASGS 2019 LGA.xlsx"
df_move = pd.read_excel(popfile,sheet_name='LGA population accounts', header=5, skipfooter=3, usecols="A:C")
df_move.columns=['LGA','pop_move','2016-2021']

df_move_melt = pd.melt(df_move,id_vars=['LGA','pop_move'], value_vars=['2016-2021'], var_name='year')
df_move_pivot = df_move_melt.pivot(index=['LGA','year'], columns='pop_move', values='value').reset_index()
df_move_pivot['LGA'] = df_move_pivot.LGA.str.split('(').str.get(0)
df_move_pivot['pop_delta'] = df_move_pivot['Population at End of Period'] - df_move_pivot['Population at Start of Period']
df_move_pivot['LGA'] = df_move_pivot['LGA'].str.strip()

df_move_pivot.head()

pop_move                LGA       year      Births      Deaths  \
0                    Albury  2016-2021      3390.0      2219.0   
1         Armidale Regional  2016-2021      1768.0      1266.0   
2                   Ballina  2016-2021      1790.0      2491.0   
3                 Balranald  2016-2021       194.0        96.0   
4         Bathurst Regional  2016-2021      2500.0      1710.0   

pop_move    Natural change    Net Migration (all sources)  \
0                   1171.0                         1031.0   
1                    502.0                         1921.0   
2                   -701.0                         1945.0   
3                     98.0                            8.0   
4                    790.0                         1277.0   

pop_move  Population at End of Period  Population at Start of Period  \
0                             54374.0                        52171.0   
1                             32736.0                        30313.0   
2                             44237.0                        42993.0   
3                              2437.0                         2330.0   
4                             44310.0                        42244.0   

pop_move  pop_delta  
0            2203.0  
1            2423.0  
2            1244.0  
3             107.0  
4            2066.0

In [46]:
move = df_move_pivot
LGA_move = list(move['LGA'].unique())
LGA_hhold = list(df_hhold_wide['LGA'].unique())

In [47]:
# Check if movement df and hhold df has the same LGAs

print("Number of LGAs in the move df:", len(LGA_move))
move_hhold = list(set(LGA_hhold).intersection(LGA_move))
print("Common LGAs in raw move and raw hhold df:", len(move_hhold))

Number of LGAs in the move df: 129
Common LGAs in raw move and raw hhold df: 0


So the same renaming procedure can be used to the move df.

In [48]:
# Match name, except for unincorporated NSW
move.loc[move.LGA=='Albury', 'LGA'] = 'ALBURY CITY'
move.loc[move.LGA=='Lithgow', 'LGA'] = 'LITHGOW CITY'
move.loc[move.LGA=='Nambucca', 'LGA'] = 'NAMBUCCA VALLEY'
move.loc[move.LGA=='Parramatta', 'LGA'] = 'CITY OF PARRAMATTA'
move.loc[move.LGA=='Upper Hunter Shire', 'LGA'] = 'UPPER HUNTER'
move.loc[move.LGA=='Warrumbungle Shire', 'LGA'] = 'WARRUMBUNGLE'

move['LGA'] = move['LGA'].str.upper()

move.head()

pop_move                LGA       year      Births      Deaths  \
0               ALBURY CITY  2016-2021      3390.0      2219.0   
1         ARMIDALE REGIONAL  2016-2021      1768.0      1266.0   
2                   BALLINA  2016-2021      1790.0      2491.0   
3                 BALRANALD  2016-2021       194.0        96.0   
4         BATHURST REGIONAL  2016-2021      2500.0      1710.0   

pop_move    Natural change    Net Migration (all sources)  \
0                   1171.0                         1031.0   
1                    502.0                         1921.0   
2                   -701.0                         1945.0   
3                     98.0                            8.0   
4                    790.0                         1277.0   

pop_move  Population at End of Period  Population at Start of Period  \
0                             54374.0                        52171.0   
1                             32736.0                        30313.0   
2                             44237.0                        42993.0   
3                              2437.0                         2330.0   
4                             44310.0                        42244.0   

pop_move  pop_delta  
0            2203.0  
1            2423.0  
2            1244.0  
3             107.0  
4            2066.0

In [49]:
# Check matching result

LGA_MOVE = list(move['LGA'].unique())
move_map= list(set(LGA_map).intersection(LGA_MOVE))
print("matched LGA in mapping and move:", len(move_map))

matched LGA in mapping and move: 128


#### Population Age
`ageband`

In [50]:
popfile = "Files/Population/2019 NSW Population Projections ASGS 2019 LGA.xlsx"
df_age = pd.read_excel(popfile,sheet_name='LGA Sex Age projections',header=5,usecols="A:E",skipfooter=3)
df_age.columns=['LGA','sex','age','2016','2021']
df_age['age_delta'] = df_age['2021'] - df_age['2016']
df_age['LGA'] = df_age.LGA.str.split('(').str.get(0)

df_age_pivot = pd.pivot_table(df_age,index=['LGA','age'], values=['2016','2021','age_delta'], 
               aggfunc=({'2016':np.sum, '2021':np.sum, 'age_delta':np.sum})).reset_index()
df_age_pivot['LGA'] = df_age_pivot['LGA'].str.strip()


LGA    age  2016  2021  age_delta
0  Albury  00-04  3505  3401       -104
1  Albury  05-09  3279  3510        231
2  Albury  10-14  3228  3370        142
3  Albury  15-19  3381  3306        -75
4  Albury  20-24  3744  3448       -296

In [72]:
#clusters

Child = df_age_pivot.age.unique()[:3]
Youth = df_age_pivot.age.unique()[3:5]
Adult = df_age_pivot.age.unique()[5:9]
MiddleAge = df_age_pivot.age.unique()[9:13]
Senior = df_age_pivot.age.unique()[13:]

print('Child',Child)
print('Youth',Youth)
print('Adult',Adult)
print('MiddleAge',MiddleAge)
print('Senior',Senior)

Child ['00-04' '05-09' '10-14']
Youth ['15-19' '20-24']
Adult ['25-29' '30-34' '35-39' '40-44']
MiddleAge ['45-49' '50-54' '55-59' '60-64']
Senior ['65-69' '70-74' '75-79' '80-84' '85+']


In [77]:
age_categ = [df_age_pivot['age'].isin(Child),
             df_age_pivot['age'].isin(Youth),
             df_age_pivot['age'].isin(Adult),
             df_age_pivot['age'].isin(MiddleAge),
             df_age_pivot['age'].isin(Senior)]
age_output = ['Child','Youth','Adult','MiddleAge','Senior']

df_age_pivot['age_bracket'] = np.select(age_categ,age_output)

age_bracket_delta = pd.pivot_table(df_age_pivot, index=['LGA'], columns=['age_bracket'], 
                                   values='age_delta', aggfunc=np.sum)
ageband = age_bracket_delta
ageband.reset_index(inplace=True)
ageband.head()

age_bracket                LGA  Adult  Child  MiddleAge  Senior  Youth
0                  ALBURY CITY    485    269        302    1519   -371
1            ARMIDALE REGIONAL   1501    195        211     630   -114
2                      BALLINA    -62   -186          5    1535    -48
3                    BALRANALD     14     36        -29     106    -20
4            BATHURST REGIONAL    716    -73        476    1062   -112

In [78]:
# Check if ageband df and hhold df has the same LGAs
LGA_age = list(ageband['LGA'].unique())

print("Number of LGAs in the ageband df:", len(LGA_age))
age_hhold = list(set(LGA_hhold).intersection(LGA_age))
print("Common LGAs in raw move and raw hhold df:", len(age_hhold))

Number of LGAs in the ageband df: 129
Common LGAs in raw move and raw hhold df: 129


In [79]:
# Apply the same renaming procedure to ageband
ageband.loc[ageband.LGA=='Albury', 'LGA'] = 'ALBURY CITY'
ageband.loc[ageband.LGA=='Lithgow', 'LGA'] = 'LITHGOW CITY'
ageband.loc[ageband.LGA=='Nambucca', 'LGA'] = 'NAMBUCCA VALLEY'
ageband.loc[ageband.LGA=='Parramatta', 'LGA'] = 'CITY OF PARRAMATTA'
ageband.loc[ageband.LGA=='Upper Hunter Shire', 'LGA'] = 'UPPER HUNTER'
ageband.loc[ageband.LGA=='Warrumbungle Shire', 'LGA'] = 'WARRUMBUNGLE'

ageband['LGA'] = ageband['LGA'].str.upper()

ageband.head()

age_bracket                LGA  Adult  Child  MiddleAge  Senior  Youth
0                  ALBURY CITY    485    269        302    1519   -371
1            ARMIDALE REGIONAL   1501    195        211     630   -114
2                      BALLINA    -62   -186          5    1535    -48
3                    BALRANALD     14     36        -29     106    -20
4            BATHURST REGIONAL    716    -73        476    1062   -112

In [53]:
LGA_AGE = list(ageband['LGA'].unique())
LGA_AGE

['ALBURY CITY',
 'ARMIDALE REGIONAL',
 'BALLINA',
 'BALRANALD',
 'BATHURST REGIONAL',
 'BAYSIDE',
 'BEGA VALLEY',
 'BELLINGEN',
 'BERRIGAN',
 'BLACKTOWN',
 'BLAND',
 'BLAYNEY',
 'BLUE MOUNTAINS',
 'BOGAN',
 'BOURKE',
 'BREWARRINA',
 'BROKEN HILL',
 'BURWOOD',
 'BYRON',
 'CABONNE',
 'CAMDEN',
 'CAMPBELLTOWN',
 'CANADA BAY',
 'CANTERBURY-BANKSTOWN',
 'CARRATHOOL',
 'CENTRAL COAST',
 'CENTRAL DARLING',
 'CESSNOCK',
 'CLARENCE VALLEY',
 'COBAR',
 'COFFS HARBOUR',
 'COOLAMON',
 'COONAMBLE',
 'COOTAMUNDRA-GUNDAGAI REGIONAL',
 'COWRA',
 'CUMBERLAND',
 'DUBBO REGIONAL',
 'DUNGOG',
 'EDWARD RIVER',
 'EUROBODALLA',
 'FAIRFIELD',
 'FEDERATION',
 'FORBES',
 'GEORGES RIVER',
 'GILGANDRA',
 'GLEN INNES SEVERN',
 'GOULBURN MULWAREE',
 'GREATER HUME SHIRE',
 'GRIFFITH',
 'GUNNEDAH',
 'GWYDIR',
 'HAWKESBURY',
 'HAY',
 'HILLTOPS',
 'HORNSBY',
 'HUNTERS HILL',
 'INNER WEST',
 'INVERELL',
 'JUNEE',
 'KEMPSEY',
 'KIAMA',
 'KU-RING-GAI',
 'KYOGLE',
 'LACHLAN',
 'LAKE MACQUARIE',
 'LANE COVE',
 'LEETON',
 'L

In [54]:
# Check matching result

age_map= list(set(LGA_map).intersection(LGA_AGE))
print("matched LGA in mapping and ageband:", len(age_map))

matched LGA in mapping and ageband: 128


## Merge the three population related data together

`hhold` `move` `ageband` 

In [55]:
hhold.head(1)

LGA  hhold_count_2016  hhold_count_2021  hhold_count_delta
0  ALBURY CITY             21940             23227               1287

In [56]:
move = move.rename(columns={'    Births':'Births_16_21', 
                            '    Deaths':'Deaths_16_21',
                            'Population at End of Period':'population_2021',
                            'Population at Start of Period':'population_2016'})
#move = move.drop(columns=['year'],axis=1)

move.head(1)

pop_move          LGA       year  Births_16_21  Deaths_16_21  \
0         ALBURY CITY  2016-2021        3390.0        2219.0   

pop_move    Natural change    Net Migration (all sources)  population_2021  \
0                   1171.0                         1031.0          54374.0   

pop_move  population_2016  pop_delta  
0                 52171.0     2203.0

In [57]:
ageband = ageband.rename(columns={'age':'age_band', 
                        '2016':'age_2016',
                        '2021':'age_2021'})
ageband.head(5)

LGA age_band  age_2016  age_2021  age_delta
0  ALBURY CITY    00-04      3505      3401       -104
1  ALBURY CITY    05-09      3279      3510        231
2  ALBURY CITY    10-14      3228      3370        142
3  ALBURY CITY    15-19      3381      3306        -75
4  ALBURY CITY    20-24      3744      3448       -296

In [59]:
pop_merging = hhold.merge(move, left_on='LGA', right_on='LGA')
pop_merged = pop_merging.merge(ageband, left_on='LGA', right_on='LGA')
pop_merged.head()

LGA  hhold_count_2016  hhold_count_2021  hhold_count_delta  \
0  ALBURY CITY             21940             23227               1287   
1  ALBURY CITY             21940             23227               1287   
2  ALBURY CITY             21940             23227               1287   
3  ALBURY CITY             21940             23227               1287   
4  ALBURY CITY             21940             23227               1287   

        year  Births_16_21  Deaths_16_21    Natural change  \
0  2016-2021        3390.0        2219.0            1171.0   
1  2016-2021        3390.0        2219.0            1171.0   
2  2016-2021        3390.0        2219.0            1171.0   
3  2016-2021        3390.0        2219.0            1171.0   
4  2016-2021        3390.0        2219.0            1171.0   

     Net Migration (all sources)  population_2021  population_2016  pop_delta  \
0                         1031.0          54374.0          52171.0     2203.0   
1                         1031.0          54374.0          52171.0     2203.0   
2                         1031.0          54374.0          52171.0     2203.0   
3                         1031.0          54374.0          52171.0     2203.0   
4                         1031.0          54374.0          52171.0     2203.0   

  age_band  age_2016  age_2021  age_delta  
0    00-04      3505      3401       -104  
1    05-09      3279      3510        231  
2    10-14      3228      3370        142  
3    15-19      3381      3306        -75  
4    20-24      3744      3448       -296

In [60]:
print("df shape:", pop_merged.shape)
print("number of LGAs in pop_merged:", pop_merged['LGA'].nunique())

df shape: (2322, 16)
number of LGAs in pop_merged: 129


In [71]:
pop_merged.to_csv('Files/Population/Population_cleaned_081021.csv', index=False)

## TESTING: Merge 3 x population DFs in to master##

In [66]:
# recall master sales/rent df w. LGA
master_map.head()

postcode             LGA  skey time_period  year quarter dwelling_type  \
0      2000          SYDNEY  s122     2017 Q3  2017      Q3         Total   
1      2000  CITY OF SYDNEY  s122     2017 Q3  2017      Q3         Total   
2      2000          SYDNEY  s122     2017 Q3  2017      Q3        Strata   
3      2000  CITY OF SYDNEY  s122     2017 Q3  2017      Q3        Strata   
4      2000          SYDNEY  s123     2017 Q4  2017      Q4         Total   

   median_price   mean_price  sales_no  ...  Qdelta_count  Adelta_count  rkey  \
0        1350.0  1516.328059     135.0  ...       -0.3250       -0.3112  r121   
1        1350.0  1516.328059     135.0  ...       -0.3250       -0.3112  r121   
2        1350.0  1516.328059     135.0  ...       -0.3182       -0.2703  r121   
3        1350.0  1516.328059     135.0  ...       -0.3182       -0.2703  r121   
4        1315.0  1823.000000     183.0  ...        0.3456       -0.0758  r122   

   median_rent_newb new_bonds_no  total_bonds_no  Qdelta_median_rent  \
0             640.0       1169.0          7914.0              -0.200   
1             640.0       1169.0          7914.0              -0.200   
2             640.0       1169.0          7914.0              -0.200   
3             640.0       1169.0          7914.0              -0.200   
4             720.0        730.0          7157.0               0.125   

   Qdelta_new_bonds  Adelta_median_rent  Adelta_new_bonds  
0            0.5545                 NaN               NaN  
1            0.5545                 NaN               NaN  
2            0.5545                 NaN               NaN  
3            0.5545                 NaN               NaN  
4           -0.3755             -0.0649           -0.0014  

[5 rows x 22 columns]

In [67]:
master_LGA = list(master_map['LGA'].unique())
pop_LGA = list(pop_merged['LGA'].unique())
shared_LGA = list(set(master_LGA).intersection(pop_LGA))
print("Number of shared LGAs in master df and population df:", len(shared_LGA))

Number of shared LGAs in master df and population df: 128


### Hence there should be 128 LGAs after merging

In [69]:
master_pop = master_map.merge(pop_merged, left_on='LGA', right_on='LGA')
master_pop.head(1)

postcode     LGA  skey time_period  year_x quarter dwelling_type  \
0      2000  SYDNEY  s122     2017 Q3    2017      Q3         Total   

   median_price   mean_price  sales_no  ...  Deaths_16_21    Natural change  \
0        1350.0  1516.328059     135.0  ...        4074.0            8288.0   

     Net Migration (all sources)  population_2021 population_2016  pop_delta  \
0                         5887.0         236891.0        222717.0    14174.0   

   age_band  age_2016  age_2021  age_delta  
0     00-04      7775     10787       3012  

[1 rows x 37 columns]

In [70]:
print("df shape:", master_pop.shape)
print("number of LGAs in master_pop:", master_pop['LGA'].nunique())

df shape: (635202, 37)
number of LGAs in master_pop: 128


In [ ]:
master_pop.to_csv('Files/Cleaned/Master_Sales_Rent_2017Q4_2021Q1_wLGA_Pop.csv', index=False)